Change event log

rtibbles opened the pull request:
## Summary
### Description of the change(s) you made
* Implements a change event log on the server side
* Makes all change application run inside an async task on the server side
* Separates out user specific and channel specific changes
* Allows syncing back of changes relevant to channels and users


### Manual verification steps performed
* Do edits
* Save edits
* Confirm they get synced
* Confirm they get applied
* Confirm they get removed from the frontend

## Reviewer guidance
### How can a reviewer test these changes?
Please test the basic flow for now.

Other thoughts:
* We could now sync task data through the repeatedly polling sync endpoint which might save us a request or two - sound worth it?
* There is some follow up that would be useful here for storing copying task info on the contentnode object to more robustly track the association
* What documentation updates would be helpful here, what is unclear in the code and could be better commented?

### Are there any risky areas that deserve extra testing?
EVERYTHING!


## References
Fixes #3034 
Fixes #2828 


## Comments
<!-- Additional comments may be added here -->

----

### Contributor's Checklist
<!-- After saving the PR, come through to tick off completed checklist items. Delete any sections that are not applicable to your PR -->

PR process:

- [ ] If this is an important user-facing change, PR or related issue the `CHANGELOG` label been added to this PR. Note: items with this label will be added to the [CHANGELOG](https://github.com/learningequality/studio/blob/master/CHANGELOG.md) at a later time
- [ ] If this includes an internal dependency change, a link to the diff is provided
- [ ] The `docs` label has been added if this introduces a change that needs to be updated in the [user docs](https://kolibri-studio.readthedocs.io/en/latest/index.html)?
- [ ] If any Python requirements have changed, the updated `requirements.txt` files also included in this PR
- [ ] Opportunities for using Google Analytics here are noted
- [ ] Migrations are [safe for a large db](https://www.braintreepayments.com/blog/safe-operations-for-high-volume-postgresql/)

Studio-specifc:

- [ ] All user-facing strings are translated properly
- [ ] The `notranslate` class been added to elements that shouldn't be translated by Google Chrome's automatic translation feature (e.g. icons, user-generated text)
- [ ] All UI components are LTR and RTL compliant
- [ ] Views are organized into `pages`, `components`, and `layouts` directories [as described in the docs](https://github.com/learningequality/studio/blob/vue-refactor/docs/architecture.md#where-does-the-frontend-code-live)
- [ ] Users' storage used is recalculated properly on any changes to main tree files
- [ ] If there new ways this uses user data that needs to be factored into our [Privacy Policy](https://github.com/learningequality/studio/tree/master/contentcuration/contentcuration/templates/policies/text), it has been noted.


Testing:

- [ ] Code is clean and well-commented
- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Any new interactions have been added to the [QA Sheet](https://docs.google.com/spreadsheets/d/1HF4Gy6rb_BLbZoNkZEWZonKFBqPyVEiQq4Ve6XgIYmQ/edit#gid=0)
- [ ] Critical and brittle code paths are covered by unit tests
___

### Reviewer's Checklist
#### This section is for reviewers to fill out.

- [ ] Automated test coverage is satisfactory
- [ ] PR is fully functional
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependency files were updated if necessary (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Contributor is in AUTHORS.md

bjester commented on contentcuration/contentcuration/models.py:
```diff
@@ -2207,3 +2211,72 @@ class Task(models.Model):
+    kwargs = JSONField(encoder=JSONEncoder)
+    applied = models.BooleanField(default=False)
+    errored = models.BooleanField(default=False)
+
```
Perhaps we should add some indices on channel and user?

bjester reviewed (commented):
Code review looks good-- left some comments. Let me know if it's ready for testing

rtibbles commented on contentcuration/contentcuration/frontend/shared/data/resources.js:
```diff
@@ -852,6 +885,36 @@ export const Session = new IndexedDBResource({
+      channelScope.id = channelId;
+      const channelRev = (window.CHANNEL_EDIT_GLOBAL || {}).channel_rev || 0;
+      await this.updateSession({
+        [`${MAX_REV_KEY}.${window.CHANNEL_EDIT_GLOBAL.channel_id}`]: channelRev,
```
No - I think this was a result of copy paste from elsewhere.

bjester commented on contentcuration/contentcuration/frontend/shared/data/serverSync.js:
```diff
@@ -79,141 +193,75 @@ function syncChanges() {
+    }
+  } catch (err) {
+    // There was an error during syncing, log, but carry on
+    console.warn('There was an error during syncing with the backend for', err); // eslint-disable-line no-console
```
This error message leaves me...

rtibbles replied:
Wanting more?

bjester commented on contentcuration/contentcuration/models.py:
```diff
@@ -2225,3 +2216,74 @@ class Task(models.Model):
+        return obj
+
+    def serialize_to_change_dict(self):
+        datum = self.kwargs
```
Shoudl this be `self.kwargs.copy()`? Otherwise, what's the reasoning I'm missing for a separate variable?

rtibbles replied:
Yes, should be a copy - as otherwise we are modifying a property of the model in place!

bjester commented on contentcuration/contentcuration/frontend/shared/data/serverSync.js:
```diff
@@ -79,141 +193,75 @@ function syncChanges() {
+    .map(([id]) => id);
+  const channel_revs = {};
+  for (let channelId of channel_ids) {
+    channel_revs[channelId] = get(user, [MAX_REV_KEY, channelId], 0);
```
Should this be 
```js
get(user, `${MAX_REV_KEY}.${channelId}`, 0)
```
???

rtibbles replied:
This is the same, the lodash `get` function allows you to specify the dot path either as a dot path or an array.

bjester replied:
But doesn't the dot path lookup follow nesting so the structure would have to be `{[MAX_REV_KEY]: { [channelId]: ... } }`? The key I see you setting is flat, not nested

bjester replied:
I guess my suggestion probably wouldn't work either with lodash `get`

rtibbles replied:
Ah, right - possible that I did not set this properly, will revist!

bjester reviewed (commented):
Left more code comments. I'll do some manual QA now

rtibbles commented on contentcuration/contentcuration/frontend/shared/data/resources.js:
```diff
@@ -852,6 +885,47 @@ export const Session = new IndexedDBResource({
+        [`${MAX_REV_KEY}.${channelId}`]: channelRev,
+      });
+      this.channelSyncKeepAlive(channelId);
+      setInterval(() => this.channelSyncKeepAlive(channelId), CHANNEL_SYNC_KEEP_ALIVE_INTERVAL);
```
Yeah, we wouldn't need to continue doing these checks if we're not online for example - but it might be more complexity than it's worth to monitor it.

rtibbles replied:
Have made some updates to at least sleep on blur.

bjester commented:
I saw this error while creating a new channel
```
django.core.exceptions.FieldError: Cannot resolve keyword 'can_edit' into field. Choices are: auth_token, bookmarked_channels, changes_about_user, changes_by_user, channel_sets, clipboard_tree, clipboard_tree_id, content_defaults, date_joined, disk_space, disk_space_used, editable_channels, email, feature_flags, files, first_name, groups, id, information, is_active, is_admin, is_staff, is_superuser, last_login, last_name, logentry, password, policies, preferences, searches, sent_by, sent_to, staged_files, task, user_permissions, view_only_channels
```

rtibbles commented:
> Seems the edit modal Saving... indicator is always visible now, even after the changes have synced.

Ah yeah, we'll need to update that check to filter out 'applied' changes.

rtibbles commented:
> I saw this error while creating a new channel

Thanks, I'll track this down.

bjester commented on contentcuration/contentcuration/viewsets/contentnode.py:
```diff
@@ -905,23 +905,43 @@ def copy(
+                        ),
+                        created_by_id=self.request.user.id,
+                        applied=True
+                    )
```
It looks like this approach will be okay for copying, but it doesn't look like it will work for publishing and may confuse devs in the future if we carry on using this pattern. The reason is that if the task requires a database transaction, then these writes for progress could be contained within the transaction and won't be readable outside of the transaction until it is committed, obviously. Looking at `map_content_nodes` that is called during publishing, it seems that it would run into this problem.

For this reason, in my task-related WIP, I utilized the task events to broadcast progress to a listener that could read and write outside of any database transactions occurring within the task.

bjester commented on contentcuration/contentcuration/viewsets/sync/endpoint.py:
```diff
@@ -3,232 +3,147 @@
+        changes_to_return = list(
+            Change.objects.filter(
+                change_filter
+            ).values("server_rev", "session_id", "channel_id", "user_id", "applied", "errored", "table", "change_type", "kwargs")
```
Does this need to be ordered?

rtibbles replied:
Possibly - I added an explicit order_by `server_rev` to be safe.

rtibbles commented on contentcuration/contentcuration/frontend/channelList/vuex/channelList/actions.js:
```diff
@@ -8,6 +8,7 @@ export function searchCatalog(context, params) {
   let promise;
   if (context.rootGetters.loggedIn) {
+    context.dispatch('channel/loadBookmarks', null, { root: true });
     promise = Channel.fetchCollection(params);
```
Yes, probably safest to do so.

rtibbles commented on contentcuration/contentcuration/frontend/shared/data/serverSync.js:
```diff
@@ -60,10 +74,124 @@ function trimChangeForSync(change) {
+    // both with any errors and the results of any merging that happened prior
+    // to the sync operation being called
+    return db[CHANGES_TABLE].where('rev')
+      .anyOf(disallowedRevs.map(Number))
```
Yes, not quite sure why I mapped twice there...

rtibbles commented on contentcuration/contentcuration/frontend/shared/data/serverSync.js:
```diff
@@ -60,10 +74,124 @@ function trimChangeForSync(change) {
+    // both with any errors and the results of any merging that happened prior
+    // to the sync operation being called
+    return db[CHANGES_TABLE].where('server_rev')
+      .anyOf(Object.keys(errorMap))
```
No, this should be ok, although I am now confused why I need the `Number` coersion at all - as both the `rev` and `server_rev` are stored on the server as Integers - so not sure why it needed to be coerced here.

rtibbles commented on contentcuration/contentcuration/frontend/shared/data/serverSync.js:
```diff
@@ -79,141 +207,76 @@ function syncChanges() {
+  }
+
+  const now = Date.now();
+  const channel_ids = Object.entries(user[ACTIVE_CHANNELS] || {})
```
Originally I was just sending `channel_ids` directly to the server, and that's why it was snake case. Now that it's being sent via `channel_revs` it's not needed. I'll update to camelCase to remove the confusion.

rtibbles commented on contentcuration/contentcuration/frontend/shared/vuex/channel/actions.js:
```diff
@@ -1,13 +1,27 @@
+      const bookmarks = await loadBookmarks(context);
+      payload.id__in = bookmarks.map(b => b.channel);
+    } else {
+      loadBookmarks(context);
```
I think I should probably just move the await above out of the if statement - the information from it is needed to correctly annotate bookmarks on the loaded channels.

rtibbles commented on contentcuration/contentcuration/frontend/shared/vuex/session/index.js:
```diff
@@ -96,10 +95,7 @@ export default {
-      });
+      Session.updateSession(currentUser);
       context.commit('ADD_SESSION', currentUser);
     },
```
Hrm, I suspect that not being await was a mistake - I'll revert that.

rtibbles commented on contentcuration/contentcuration/tasks.py:
```diff
@@ -43,154 +35,26 @@
+    changes_qs = Change.objects.filter(applied=False, errored=False, channel_id=channel_id)
+    apply_changes(changes_qs)
+    if changes_qs.exists():
+        user = User.objects.filter(editable_channels=channel_id).first()
```
Flagging that we chatted about this, and resolved to clear in follow up work.

rtibbles commented on contentcuration/contentcuration/frontend/channelEdit/vuex/contentNode/actions.js:
```diff
@@ -442,30 +441,40 @@ export function loadNodeDetails(context, nodeId) {
+      return true;
+    }
+  }
+  return false;
```
I think I've found something even more concise!

micahscopes commented:
I enjoy seeing activity on this one :man_dancing:

bjester reviewed (changes requested):
After publishing, the publish button became enabled again, even though no changes were made. New nodes that were unpublished are still marked as unpublished after publishing.

marcellamaki reviewed (changes requested):
On the UI side, there is some syncing and display weirdness that is showing up most notably with the new metadata fields, both on folders and on resource editing. If you are editing quickly (it seems like... moving between fields before a sync completes?) there is a mismatch that takes some time to resolve, and sometimes resolves in really unexpected ways. I _think_ this is new to these changes, as I've been working from unstable for the metadata display in the last few days and haven't encountered this. See the videos below. I included the network tab in the second video in case it is helpful -- there does seem to be some correlation with when `/sync` runs and when the UI changes.

https://user-images.githubusercontent.com/17235236/173371542-5bd2f3b2-f02a-4800-be23-b92065d24f5e.mp4

https://user-images.githubusercontent.com/17235236/173371554-258521d5-0c77-4417-bdab-7cdbda69aa94.mp4


The other thing I noticed which is not on hotfixes, although I can't fully tell if it's related to these changes, is the content library is not actually displaying the resources, just the number of results. For example, even though I'm running locally and there is only the "published channel" to test with, nothing appears: 

<img width="974" alt="Screen Shot 2022-06-13 at 10 01 07 AM" src="https://user-images.githubusercontent.com/17235236/173372080-3b9feea0-916e-4156-a200-b045e8356698.png">

rtibbles commented:
> After publishing, the publish button became enabled again, even though no changes were made. New nodes that were unpublished are still marked as unpublished after publishing.

The publish button should be properly toggling now, and all content nodes in the frontend are properly flagged as published and unchanged.

I did this by switching to tracking for edits since the last publish (except in the case of ricecooker channels, where I fall back to the changed attribute).

> the new metadata fields, both on folders and on resource editing

This was very reproducible and was a result of my poor decision to use dot paths in the materialized paths of our nested metadata labels. This confuses Dexie which also uses dot paths, but for describing nested keyPaths for changes. I have monkey patched Dexie to add specific excludes for our metadata labels to prevent this from causing an issue.

> I can't fully tell if it's related to these changes, is the content library is not actually displaying the resources, just the number of results

This was a bug I introduced while refactoring the bookmark code - there was an error in the `getChannels` getter function - which I have now corrected.

In addition, I have done some testing of this across two browsers, and it seems like the change application logic across browsers is working as intended - I did this after the issues that @marcellamaki had reported, to ensure that the fix I applied to Dexie was universal. I retested this after the upgrade to Dexie 3.2 and the fix persisted.

bjester reviewed (approved):
Calling it good!
